### ISSUE TITLE
IOS release stuck on splash screen if no internet connection

### ISSUE THREAD

**Issue by KristinaHus** [expo/expo#13045]
### Summary

App stuck on splash screen if I turn on airplane mode and start the app. And it won't start even if I close the app and open it again. It starts only if I kill the app from the tray and then start again.

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

iOS

### SDK Version (managed workflow only)

40.0.0

### Environment

Expo CLI 4.4.3 environment info:
    System:
      OS: macOS 11.2.2
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 12.13.1 - /usr/local/bin/node
      Yarn: 1.22.10 - /usr/local/bin/yarn
      npm: 6.12.1 - /usr/local/bin/npm
      Watchman: 4.9.0 - /opt/homebrew/bin/watchman
    Managers:
      CocoaPods: 1.10.1 - /opt/homebrew/bin/pod
    SDKs:
      iOS SDK:
        Platforms: iOS 14.5, DriverKit 20.4, macOS 11.3, tvOS 14.5, watchOS 7.4
    IDEs:
      Android Studio: 4.1 AI-201.8743.12.41.7199119
      Xcode: 12.5/12E262 - /usr/bin/xcodebuild
    npmPackages:
      expo: ^40.0.0 => 40.0.1 
      react: 16.13.1 => 16.13.1 
      react-dom: 16.13.1 => 16.13.1 
      react-native: https://github.com/expo/react-native/archive/sdk-40.0.1.tar.gz => 0.63.2 
      react-native-web: ~0.13.12 => 0.13.18 
    npmGlobalPackages:
      expo-cli: 4.4.3
    Expo Workflow: managed

### Reproducible demo or steps to reproduce from a blank project

Can't share a reproducible demo as it can be reproduced only in the release build installed from TestFlight.

Here is what I can share.
I've created an app. It uses NetInfo module it is listening for network changes in App.tsx file. I show the modal that overlays the whole app when there is no internet connection and hide it when NetInfo notifies me that there is a connection.

Next, when I turn on airplane mode and start the app I can see my popup over the Splash screen, and when I turn off the airplaine-mode the modal is gone, but the splash screen is still present and I never see the Home Screen. To make the app work again I have to close the app and remove it from the tray.

I've read that there might be a problem with OTA Updates, so I've tried to set 'updates.fallbackToCacheTimeout' to 0 or 1. Even tried 'updates.enabled' false. That didn't help. 

I've checked a device logs and the last warning I see, after enabling the network is 'nehelper sent invalid result code [1] for Wi-Fi information request'

**Comment by brentvatne** [comment]
> Can't share a reproducible demo as it can be reproduced only in the release build installed from TestFlight.

you should be able to reproduce it in a simulator too - `expo build:ios -t simulator`

i think it's possible that you have some code that runs before hiding the splash screen that doesn't handle no network connection. do you use `SplashScreen.preventAutoHideAsync()` at all?

**Comment by zholmes1** [comment]
Are you using `AppLoading` or `SplashScreen`?

If you are using `AppLoading` either your `startAsync` callback is hanging, or your `onFinish` or `onError` don't set the proper condition to bypass your `AppLoading`. For instance, let's say you have a `const [ready, setReady] = useState(false)`, and then `if (!ready) return <AppLoading...`. Both your `onFinish` and `onError` will need to result in `setReady(true)`.

If you are using `SplashScreen`, then it's a little more freeform. But you need to ensure that `hideAsync()` is called at some point.

**Comment by KristinaHus** [comment]
@brentvatne, @zholmes1 Thank you for the quick response. Yes, I use AppLoading and my first thought was that the problem is there. But it is not. I was able to figure out what is the problem. Don't think that this is expected behavior as well.
I've found that splash screen stuck when I have a Modal shown over it.

Here is how my App.tsx render method looks like 
`<ActionSheetProvider>
      <Provider store={store}>
        <PersistGate loading={null} persistor={persistor}>
          <ThemeProvider theme={{ ...theme, colors: { ...theme.colors, ...colors } }}>
            <RootStack />
            <Modal animationType="slide" visible={!hasConnection}>
              <NoNetwork onHide={onHide} />
            </Modal>
          </ThemeProvider>
        </PersistGate>
      </Provider>
    </ActionSheetProvider>`

My first thought was that I need to move the Modal to RootStack navigation. But it didn't help. Anyway, it shows up over the splash screen and blocks it.

What is interesting. Moving to RoorStack navigation fixed the issue for the simulator build. But release build downloaded from TestFlight still showing Modal over the splash screen.
